net processing: Move remaining globals into PeerManagerImpl

dergoegge opened the pull request:
This PR moves the remaining net processing globals into `PeerManagerImpl`. This will make testing the peer manager in isolation easier and also acts as a code clean up.

dergoegge committed:
[net processing] Move mapNodeState into PeerManagerImpl

dergoegge committed:
[net processing] Move block cache state into PeerManagerImpl

dergoegge committed:
scripted-diff: Rename PeerManagerImpl members

-BEGIN VERIFY SCRIPT-
ren() { sed -i "s:\<$1\>:$2:g" $(git grep -l "\<$1\>" ./src ./test); }

ren mapNodeState                              m_node_states
ren cs_most_recent_block                      m_most_recent_block_mutex
ren most_recent_block                         m_most_recent_block
ren most_recent_compact_block                 m_most_recent_compact_block
ren most_recent_block_hash                    m_most_recent_block_hash
ren fWitnessesPresentInMostRecentCompactBlock m_most_recent_compact_block_has_witnesses
ren nPreferredDownload                        m_num_preferred_download_peers
ren nHighestFastAnnounce                       m_highest_fast_announce
-END VERIFY SCRIPT-

w0xlt commented:
Concept ACK.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1747,7 +1766,7 @@ void PeerManagerImpl::RelayTransaction(const uint256& txid, const uint256& wtxid
 
 void PeerManagerImpl::_RelayTransaction(const uint256& txid, const uint256& wtxid)
 {
-    m_connman.ForEachNode([&txid, &wtxid](CNode* pnode) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
```
Note that this call to `ForEachNode()` is removed in #21160 (not suggesting you do anything differently here).

jnewbery reviewed (commented):
Concept and approach ACK. A few minor suggestions inline.

dergoegge committed:
[net processing] Move CNodeState declaration above PeerManagerImpl

dergoegge commented:
Thanks @jnewbery for the review! I took all your suggestions.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2685,8 +2686,12 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-        LOCK(cs_main);
-        UpdatePreferredDownload(pfrom, State(pfrom.GetId()));
+            LOCK(cs_main);
+            CNodeState* state = State(pfrom.GetId());
```
It may be better to move this code inline in one commit, separate from changing `nPreferredDownload` from a global to a member variable, so you're not mixing move-only changes and refactor changes in the same commit.

This code could also be cleaned up in a couple of ways now that it's inline (but that doesn't need to happen in this PR):

- `m_num_preferred_download_peers -= state->fPreferredDownload;` can be removed, since `version` processing can only happen once for each peer, and `fPreferredDownload` is initialized to false, so this line is always a no-op.
- `// Whether this node should be marked as a preferred download node.` is redundant with the `// Potentially mark this peer as a preferred download peer.` comment above and can be removed.

dergoegge replied:
Good idea, I split the inlining and clean ups into a separate commit.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -676,121 +797,19 @@ class PeerManagerImpl final : public PeerManager
-static std::map<NodeId, CNodeState> mapNodeState GUARDED_BY(cs_main);
+CNodeState* PeerManagerImpl::State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    std::map<NodeId, CNodeState>::iterator it = m_node_states.find(pnode);
```
To avoid the code duplication here, you can define this in terms of the `State() const` function, and cast away constness:

```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 877bca4c0c..6cadef2a37 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -555,8 +555,10 @@ private:
     /** Map maintaining per-node state. */
     std::map<NodeId, CNodeState> m_node_states GUARDED_BY(cs_main);
 
-    CNodeState* State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    /** Get a pointer to a const CNodeState, used when not mutating the CNodeState object.*/
     const CNodeState* State(NodeId pnode) const EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    /** Get a pointer to a mutable CNodeState.*/
+    CNodeState* State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
     uint32_t GetFetchFlags(const CNode& pfrom) const EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
@@ -798,20 +800,17 @@ private:
     bool SetupAddressRelay(const CNode& node, Peer& peer);
 };
 
-CNodeState* PeerManagerImpl::State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+const CNodeState* PeerManagerImpl::State(NodeId pnode) const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
 {
-    std::map<NodeId, CNodeState>::iterator it = m_node_states.find(pnode);
+    std::map<NodeId, CNodeState>::const_iterator it = m_node_states.find(pnode);
     if (it == m_node_states.end())
         return nullptr;
     return &it->second;
 }
 
-const CNodeState* PeerManagerImpl::State(NodeId pnode) const EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+CNodeState* PeerManagerImpl::State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
 {
-    std::map<NodeId, CNodeState>::const_iterator it = m_node_states.find(pnode);
-    if (it == m_node_states.end())
-        return nullptr;
-    return &it->second;
+    return const_cast<CNodeState*>(std::as_const(*this).State(pnode));
 }
```

See https://stackoverflow.com/a/47369227/933705. Generally const_cast is discouraged, but this is one instance where it's considered acceptable.

jnewbery reviewed (commented):
Code review ACK 8c08e33f678bd10254fc4ca1ed6f47ad52b027a5.

A couple of small suggestions inline.

dergoegge committed:
[net processing] Move nPreferredDownload into PeerManagerImpl

dergoegge committed:
[net processing] Inline and simplify UpdatePreferredDownload

We inline `UpdatePreferredDownload` because it is only used in one
location during the version handshake. We simplify it by removing the
initial subtraction of `state->fPreferredDownload` from
`nPreferredDownload`. This is ok since the version handshake is only
called once per peer and `state->fPreferredDownload` will always be
false before the newly inlined code is called, making the subtraction a
noop.

jnewbery commented on src/test/denialofservice_tests.cpp:
```diff
@@ -197,7 +197,7 @@ BOOST_AUTO_TEST_CASE(stale_tip_peer_management)
     // Update the last announced block time for the last
     // peer, and check that the next newest node gets evicted.
-    UpdateLastBlockAnnounceTime(vNodes.back()->GetId(), GetTime());
+    peerLogic->UpdateLastBlockAnnounceTime(vNodes.back()->GetId(), GetTime());
```
You can remove the `UpdateLastBlockAnnounceTime()` declaration on line 39 (it's now no longer used).

dergoegge replied:
Done, good catch!

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2685,8 +2685,10 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-        LOCK(cs_main);
-        UpdatePreferredDownload(pfrom, State(pfrom.GetId()));
+            LOCK(cs_main);
+            CNodeState* state = State(pfrom.GetId());
```
I think the commit _[net processing] Inline and simplify UpdatePreferredDownload_ could have slightly more explanation in the commit log (explaining that this logic can only be hit once per peer, so `fPreferredDownload` will always be `false`, and the `m_num_preferred_download_peers -= state->fPreferredDownload;` line is always a no-op).

jnewbery reviewed (commented):
Code review ACK 822152c50654228a926ec56819878a5429d75c2e

dergoegge commented:
Rebased

jnewbery commented:
ACK 45f3b1dbebf799b386907de1a4eed20777e92073

Verified rebase by doing it myself and comparing.

jnewbery commented:
Code review ACK 79f6a554354fa923167979faf8cf8f3120fc9bfa

I noticed the following line in `PeerManagerImpl::NewPOWValidBlock()`:

```
    static int nHighestFastAnnounce = 0;
```

It's perhaps out of scope for this PR, but perhaps that should be made a member variable of `PeerManagerImpl` alongside the cached recent block members.

dergoegge committed:
[net processing] Move nHighestFastAnnounce into PeerManagerImpl

dergoegge commented:
Rebased and added a commit (da9e51a1fe704d4a6a93de34e6837fab37655860) to move `nHighestFastAnnounce` into `PeerManagerImpl`.

jnewbery reviewed (commented):
Code review ACK 8fa7fa0cb04661c06fc1fd44667bfebcc6716a9d

dergoegge commented on src/net_processing.cpp:
```diff
@@ -567,6 +679,20 @@ class PeerManagerImpl final : public PeerManager
+    bool m_most_recent_compact_block_has_witnesses GUARDED_BY(m_most_recent_block_mutex){false};
+
+    /**
+     * Height of the highest fast announced block.
```
Good point, adjusted the comment

jnewbery commented:
Code review ACK 778343a379026ef233dffea67f5226565f6d5720

maflcko reviewed (approved):
ACK 778343a379026ef233dffea67f5226565f6d5720 🗒

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 778343a379026ef233dffea67f5226565f6d5720 🗒
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjBIwv+MjgYz0poUymP0rv825e0CNNMrgOXg5hJmGXAF7dDpzn6Kqje2icEX1Iy
vNbpPsRDachm/67cOdTJinNgw8TStVI+KxURKCSCl1A30Xz01groLbTyXjVn2gwy
Bk9HoIRpa0CLbvGenk2OW6jjxYFgbK81TKDTHqHEM6laayMlKHFaloO/AlB+SfYL
TRNCBjxGnzoBzPfG3d5JMI5Rt15lBSvKRm6dCe5pnUdDLJyfJLVT5CIb606Tkxoz
0ASF4H0nvsdPCR5t3T+spy8fFDgmsUwtUNbpOAKKkm3Vmr1EGxHxyGCW3Uyq3eLQ
YKefQrFjp21dcxeCeCcsZp30Ropssp1Dc1bOBgPIworvUzTlb8H/WU1aDr16Eis+
/GITCLQ75eifYYaPPEPnJxLwW8QLE4k38zhmYKtR8EafRpB8Zh1NAt+TSWB8iSOl
yjRfMQGD7TeIDVnXzOz+fEwzI7JVsexqhu1BD14Xfx3pUsXfclbnFWHH2/0XaTwp
59A5jsEa
=56fQ
-----END PGP SIGNATURE-----
```


</details>

Meru852 commented:
> ACK [778343a](https://github.com/bitcoin/bitcoin/commit/778343a379026ef233dffea67f5226565f6d5720) 🗒
> 
> Show signature

Marco i suggest you read your tele first. do you still in freebitco.dev isn it? 
reply me asap.
